doc/architecture/decisions/0015-key-management.md

```diff
@@ -0,0 +1,222 @@
+
+```rust
+// generate new random seed hierarchy
+let hd = DpkiHd::new_random();
+
+// display the primary/root seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_primary_seed()));
+
+// attach to existing identitiy
+let existing_app_pub_data = DpkiApp::fetch(
+    prompt_user_for_dpki_id());
+existing_app_pub_data.request_authorize_new_keys(hd);
+// user will need to sign the new keys into their chain from
+// another device that is already authorized
+
+// initialize a device structure with device seed number 1
+let device = DpkiDevice::new_from_seed(hd.get_device_seed(1));
+// (optional) display device seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_device_seed(1)));
+// blink the device pin on the holoport circle
+display_to_user(device.get_device_pin());
+
+// reboot will require passphrase re-entry
+save_to_disk(device.get_saved(prompt_user_for_passphrase()));
+```
+
+##### 5 - new holo identity
+
+```rust
+// will generate a new random salt and password hash it
+// to form a seed for keypair generation
+let holo = DpkiHolo::new_random_from_passphrase(
+    prompt_user_for_passphrase());
+
+// publish on dpki dht
+let app_pub_data = DpkiApp::new_holo(
+    prompt_user_for_username(),
+    holo);
+publish_to_dpki_dht(app_pub_data);
+
+// init a device object for signing / encryption
+let device = DpkiDevice::new_from_holo(holo);
+```
+
+##### 6 - restore holo identity
+
+```rust
+// we need to fetch the distributed salt
+// in order to restore our private keys
+let username = prompt_user_for_username();
+let app_pub_data = DpkiApp::fetch(
+    lookup_dpki_id(username));
+
+// generate the same seed data / private keys from salt / passphrase
+let holo = DpkiHolo::new_from_salt_passphrase(
+    app_pub_data.get_salt_for_username(username),
+    prompt_user_for_passphrase());
+
+// init a device object for signing / encryption
+let device = DpkiDevice::new_from_holo(holo);
+```
+
+## Decision
+
+We will use the [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) standard for generating mnemonic word lists from seeds  [Rust crate is available.](https://github.com/infincia/bip39-rs)
+
+We will use [libsodium](https://download.libsodium.org/doc/) as our primary library for encryption, decryption, signatures, password hashing, etc..
+
+We will use this api sketch as a starting point for implementation in the `holochain_api` crate, i.e. for use by container/composer developers:
+
+```rust
+/// utilities for working with mnemonic word lists
+pub mod Dpki {
+    fn seed_to_mnemonic(seed: &[u8; 32]) -> String;
+    fn mnemonic_to_seed(mnemonic: &str) -> [u8; 32];
+}
+
+/// struct for hd seed derivation handling
+impl DpkiHd {
+    pub fn new_random() -> Self;
+    pub fn new_from_primary_seed(primary_seed: &[u8; 32]) -> Self;
+    pub fn get_primary_seed(&self) -> [u8; 32];
+    pub fn get_revoke_seed(&self) -> [u8; 32];
```

Comment:
I get confused by these sub-keys also being called `seed`. Maybe I'm misunderstanding - we are talking about keys in all instances, right? The root key that is used to create the sub-keys (in a deterministic manner) is called the *seed* because it is what other keys are derived from via the derivation paths listed below. But then the keys that are created this way are keys, **not seeds.** Right?

Reply:
Seeds are not keys.  Seeds are 32bit numbers from which you can create an actual private key, using what ever algorithm we use for keys be that signing keys (network/or authoring) or encryption keys.  You can't use a seed as an actual key, i.e. it's not a value you can pass into the code library as a key.  Its a value you use to create the key.  But in some sense you can think of it as a key because it deterministically generates one.  So you get hierarchical deterministic keys because of the hierarchy of seeds.

Reply:
Right. So why are the names of this functions get_*_seed(). What they return is a derived thing, which is a key. No?

Reply:
LibSodium lets you use 32 byte (not bit ... 256 bits) seeds as well as 32 byte keys. However, a seed is not a key, it is the source of entropy for generating key pairs.  I believe a seed doesn't have its counterpart pub/private pairs. It is used with an (elliptical curve) algorithm for selecting the points on the curve which give you a pairing.

Reply:
Ok, so you are both confirming that key != seed.
Isn't the names of these functions misleading then? They are creating something that is derived from a seed. So it must be key.

Or are you saying, @artbrock, that since both are 32bit we just use the generated key as a seed to introduce another level in the hierarchy? Is that how you do that? I don't know if I'm asking a dumb question or if I spotted a mistake.

Reply:
In libsodium, derivation takes an input seed buffer and produces another seed buffer in a deterministic and non-reversible manner. This can be done as many times as desired.

Any of those seed buffers can be used to produce asymmetric keypairs for signing or encryption. The same seed will always produce the same keypair, making them equivalent from a security exposure perspective to a private key.

The suggested api works with the seeds for simplicity. DpkiDevice and DpkiApp will use the seed to generate the appropriate keypairs for use in encryption and signing.